extlib/millhone/src/api.rs

```diff
@@ -38,24 +50,308 @@ pub mod prelude {
+            .target(snippets::Target::Function.to_string())
+            .kind(snippet.metadata().kind().to_string())
+            .method(snippet.metadata().method().to_string())
+            .file_path(path.to_string_lossy().to_string())
+            .byte_start(location.start_byte() as _)
+            .byte_end(location.end_byte() as _)
+            .line_start(text_loc.line_start as _)
+            .line_end(text_loc.line_end as _)
+            .col_start(text_loc.col_start as _)
+            .col_end(text_loc.col_end as _)
+            .language(L::display().to_string())
+            .ingest_id(ingest_id.to_string())
+            .build()
+    }
+}
+
+/// A deterministic representation of source code.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, From)]
+pub struct Fingerprint(Vec<u8>);
+
+impl Fingerprint {
+    /// Read the fingerprint as a byte vec.
+    pub fn to_vec(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+impl Serialize for Fingerprint {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let encoded = BASE64_STANDARD.encode(&self.0);
+        serializer.serialize_str(&encoded)
+    }
+}
+
+impl<'de> Deserialize<'de> for Fingerprint {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw = String::deserialize(deserializer)?;
+        let decoded = BASE64_STANDARD
+            .decode(raw)
+            .map_err(serde::de::Error::custom)?;
+        Self(decoded).pipe(Ok)
+    }
+}
+
+fn build_default_agent(creds: Option<Credentials>) -> Agent {
     let app_name = env!("CARGO_PKG_NAME");
     let app_version = env!("CARGO_PKG_VERSION");
-    AgentBuilder::new()
+
+    let mut builder = AgentBuilder::new()
         // The intention is to provide the service with the name and version
         // so that we can track deployed app versions over time.
         .user_agent(&format!("{app_name}/{app_version}"))
         // Not based on anything specific but seems reasonable.
-        .timeout(Duration::from_secs(30))
-        .build()
+        .timeout(Duration::from_secs(30));
+
+    // Inject authorization into every request here so that each method doesn't have to remember to.
+    if let Some(creds) = creds {
+        let credentials_header = creds.as_basic();
+        builder = builder.middleware(move |req: Request, next: MiddlewareNext<'_>| {
+            next.handle(req.set("Authorization", &credentials_header))
+        })
+    }
+
+    builder.build()
+}
+
+/// The location, using textual "line, column" indicators, for a snippet.
+///
+/// Line end is inclusive: if a snippet starts on line 1 and ends before
+/// the end of the line, `line_end` is line 1 as well.
+///
+/// Column end is inclusive: when displaying the snippet, the column
+/// _indicated by_ this number should be included.
+///
+/// For example, if the full text was "hello world!",
```

Comment:
Are bytes zero-indexed and but columns and lines 1-indexed? I think that makes sense, but that should be noted in this comment

Reply:
Maybe also do a bit of a text representation here, like you do in the tests:

```
         hello world!
  bytes: 012345678901
               ^   ^
columns: 123456789012
               ^   ^
```

Reply:
I'm revewing the next PR, and it looks like this gets deleted in https://github.com/fossas/fossa-cli/pull/1280/files, so feel free to ignore these comments :)